### PR TITLE
[CosmosDB] Adding apiVersion 2021-07-01-preview

### PR DESCRIPTION
**Pull request by ajparule**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [x] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? July 2021
  3. When you expect to publish swagger? July 2021
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  -  [ ] Ensure to copy the existing version into new directory structure for first commit (including refactoring) and then push new changes including version updates in separate commits. This is required to review the changes efficiently.
  - Adding a new service

- [ ] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by RamyasreeChakka** [comment]
>             "description": "The graph create or update operation will complete asynchronously."

Please use 201 + ProvisioningState pattern for asyns PUT.
Refer RPC guidelines here [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#creatingupdating-using-put](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#creatingupdating-using-put)

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-07-01-preview/cosmos-db.json:1557 in c782265. [](commit_id = c7822654639aee82432eb831114a0e0dcb660109, deletion_comment = False)

**Comment by RamyasreeChakka** [comment]
>             "description": "The Cassandra View create or update operation will complete asynchronously."

For PUT operations, use  201 + ProvisioningState pattern. Refer RPC guidelines here [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#creatingupdating-using-put](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#creatingupdating-using-put)

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-07-01-preview/cosmos-db.json:5395 in c782265. [](commit_id = c7822654639aee82432eb831114a0e0dcb660109, deletion_comment = False)

**Comment by RamyasreeChakka** [comment]
>       "get": {

Please add a GET collection API on throughputSettings.
Any resource type MUST implement GET for collection and single resource regardless if it's a singleton or not.

Refer best practice 'Inconsistent or incomplete CRUD' here [https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html](https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html)

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-07-01-preview/cosmos-db.json:5448 in c782265. [](commit_id = c7822654639aee82432eb831114a0e0dcb660109, deletion_comment = False)

**Comment by RamyasreeChakka** [comment]
>         "value": {

Define nextLink property too.

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-07-01-preview/cosmos-db.json:5822 in c782265. [](commit_id = c7822654639aee82432eb831114a0e0dcb660109, deletion_comment = False)

**Comment by RamyasreeChakka** [comment]
>           "description": "Opt-out of local authentication and ensure only MSI and AAD can be used exclusively for authentication.",

Avoid boolean properties. Use enum localAuth with values enabled/disabled.
Refer best practice 'Replace boolean/switch properties with better enum' here [https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html](https://armwiki.azurewebsites.net/rp_onboarding/process/api_review_best_practices.html)

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-07-01-preview/cosmos-db.json:6864 in c782265. [](commit_id = c7822654639aee82432eb831114a0e0dcb660109, deletion_comment = False)

**Comment by MehaKaushik** [comment]
>             "description": "The graph create or update operation will complete asynchronously."

This is not a tracked resource, it is a proxy resource. There is no provisioning state defined for it. 

---
In reply to: [912766213](https://github.com/Azure/azure-rest-api-specs/pull/14817#issuecomment-912766213) [](http://example.com/codeflow?ancestors=912766213,912766213)

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-07-01-preview/cosmos-db.json:1557 in c782265. [](commit_id = c7822654639aee82432eb831114a0e0dcb660109, deletion_comment = False)

**Comment by MehaKaushik** [comment]
>             "description": "The Cassandra View create or update operation will complete asynchronously."

This is not a tracked resource, it is a proxy resource. There is no provisioning state defined for it. 

---
In reply to: [912767501](https://github.com/Azure/azure-rest-api-specs/pull/14817#issuecomment-912767501) [](http://example.com/codeflow?ancestors=912767501,912767501)

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-07-01-preview/cosmos-db.json:5395 in c782265. [](commit_id = c7822654639aee82432eb831114a0e0dcb660109, deletion_comment = False)

**Comment by MehaKaushik** [comment]
>       "get": {

Point noted, we do not have the get list API ready and functional in our service at the moment. We will add them soon to complete this. Can you please not block this PR for this reason.

---
In reply to: [912769031](https://github.com/Azure/azure-rest-api-specs/pull/14817#issuecomment-912769031) [](http://example.com/codeflow?ancestors=912769031,912769031)

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-07-01-preview/cosmos-db.json:5448 in c782265. [](commit_id = c7822654639aee82432eb831114a0e0dcb660109, deletion_comment = False)

**Comment by MehaKaushik** [comment]
>           "description": "Opt-out of local authentication and ensure only MSI and AAD can be used exclusively for authentication.",

This is not a new property being added in this apiVersion and was reviewed when originally added. We cannot change it now. 
Other services which have disableLocalAuth as booleans: 
[https://docs.microsoft.com/en-us/azure/azure-monitor/logs/azure-ad-authentication-logs](https://docs.microsoft.com/en-us/azure/azure-monitor/logs/azure-ad-authentication-logs)

[https://docs.microsoft.com/en-us/azure/templates/microsoft.machinelearningservices/workspaces/computes?tabs=json](https://docs.microsoft.com/en-us/azure/templates/microsoft.machinelearningservices/workspaces/computes?tabs=json)
 

---
In reply to: [912771375](https://github.com/Azure/azure-rest-api-specs/pull/14817#issuecomment-912771375) [](http://example.com/codeflow?ancestors=912771375,912771375)

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-07-01-preview/cosmos-db.json:6864 in c782265. [](commit_id = c7822654639aee82432eb831114a0e0dcb660109, deletion_comment = False)

**Comment by majastrz** [comment]
> Added some comments and/or questions. Please take a look.

I added this comment to the wrong PR. Will resume reviewing...

**Comment by majastrz** [comment]
>         "responses": {

You marked the delete as long running but 202 is missing from the responses.

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-07-01-preview/cosmos-db.json:1593 in 6acf584. [](commit_id = 6acf5842c8a0078a03b177be1e38f290b21a1d36, deletion_comment = False)

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-07-01-preview/restorable.json [hunk 1]
```diff
@@ -0,0 +1,1120 @@
...
+          },
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "cosmos-db.json#/definitions/CloudError"
+            }
+          }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": null
```
Comment (majastrz) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-07-01-preview/restorable.json:72]: >null [](http://example.com/codeflow?start=26&length=4)

Are you really not going to support paging on all the new list calls? Adding paging in the future is a breaking change.

Keep in mind that you can declaring paging in the swagger but still have the backend service always return 1 page.

**Review by majastrz — APPROVED** [review]
Signed off from ARM side with some minor comments and/or questions.

**Comment by MehaKaushik** [comment]
>         "responses": {

Done

---
In reply to: [914862326](https://github.com/Azure/azure-rest-api-specs/pull/14817#issuecomment-914862326) [](http://example.com/codeflow?ancestors=914862326,914862326)

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-07-01-preview/cosmos-db.json:1593 in 6acf584. [](commit_id = 6acf5842c8a0078a03b177be1e38f290b21a1d36, deletion_comment = False)

**Review by MehaKaushik — APPROVED** [review]

**Comment by ajparule** [comment]
>         "value": {

By Design

---
In reply to: [912769887](https://github.com/Azure/azure-rest-api-specs/pull/14817#issuecomment-912769887) [](http://example.com/codeflow?ancestors=912769887,912769887)

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-07-01-preview/cosmos-db.json:5822 in 12ab9a3. [](commit_id = 12ab9a331a2c3472d72f79e4115ff6b40f5f1bcf, deletion_comment = False)

**Comment by leni-msft** [comment]
@ArcturusZhang please help check the go sdk generation failure.

**Comment by leni-msft** [comment]
@ajparule @MehaKaushik is the service ready for the new apiVersion?

**Comment by leni-msft** [comment]
@msyyc please help check the python sdk breaking changes.

**Comment by MehaKaushik** [comment]
> @ajparule @MehaKaushik is the service ready for the new apiVersion?

yes it is

**Comment by MehaKaushik** [comment]
> @ArcturusZhang please help check the go sdk generation failure.

The breaking changes are because of differences in preview and stable apiVersion. Going forward our team plans to release only preview apiVersion SDKs, that will prevent such breaking cases. The same holds for Go Sdk.

File: specification/cosmos-db/resource-manager/readme.go.md [hunk 2]
```diff
@@ -21,6 +21,7 @@ azure-arm: true
 
 ``` yaml $(go) && $(multiapi)
 batch:
+  - tag: package-2021-07-preview
```
Comment (ArcturusZhang) [inline specification/cosmos-db/resource-manager/readme.go.md:24]: Please assign a `output-folder` for this tag. You can take the entries below as reference

**Review by ArcturusZhang — CHANGES_REQUESTED** [review]

File: specification/cosmos-db/resource-manager/readme.go.md (hunk 2 above)
Comment (MehaKaushik) [reply specification/cosmos-db/resource-manager/readme.go.md:24]: Done

**Review by ArcturusZhang — APPROVED** [review]

**Comment by msyyc** [comment]
approve for python breaking change

**Pull request merged**